packages/flask-aserto/src/flask_aserto/middleware.py

```diff
@@ -0,0 +1,215 @@
+        policy_path = self._policy_path_resolver()
+        
+        decisions = client.decisions(
+            policy_path=policy_path,
+            decisions=(decision,),
+            policy_instance_name=self._policy_instance_name,
+            policy_instance_label=self._policy_instance_label,
+            resource_context=resource_context,
+        )
+        return decisions[decision]
+
+    @overload
+    def authorize(self, handler: Handler) -> Handler:
+        ...
+
+    @overload
+    def authorize(
+        self,
+        *,
+        authorizer_options: AuthorizerOptions = ...,
+        identity_provider: IdentityMapper = ...,
+        policy_instance_name: str = ...,
+        policy_instance_label: str = ...,
+        policy_path_root: str = ...,
+        policy_path_resolver: StringMapper = ...,
+    ) -> Callable[[Handler], Handler]:
+        ...
+
+    def authorize(
+        self,
+        *args: Any,
+        **kwargs: Any,
+    ) -> Union[Handler, Callable[[Handler], Handler]]:
+        arguments_error = TypeError(
+            f"{self.authorize.__name__}() expects either exactly 1 callable"
+            " 'handler' argument or at least 1 'options' argument"
+        )
+
+        handler: Optional[Handler] = None
+
+        if not args and kwargs.keys() == {"handler"}:
+            handler = kwargs["handler"]
+        elif not kwargs and len(args) == 1:
+            (handler,) = args
+
+        if handler is not None:
+            if not callable(handler):
+                raise arguments_error
+            return self._authorize(handler)
+
+        if args:
+            raise arguments_error
+
+        return self._with_overrides(**kwargs)._authorize
+
+    def _authorize(self, handler: Handler) -> Handler:
+        if self._policy_instance_name == None:
+            raise TypeError(f"{self._policy_instance_name}() should not be None")
+        
+        if self._policy_instance_label == None:
+            self._policy_instance_label = self._policy_instance_name
+
+        @wraps(handler)
+        def decorated(*args: Any, **kwargs: Any) -> Response:
+            client = self._generate_client()
+            resource_context = self._resource_context_provider()
+            policy_path = self._policy_path_resolver()
+
+            decisions = client.decisions(
+                policy_path=policy_path,
+                decisions=("allowed",),
+                policy_instance_name=self._policy_instance_name,
+                policy_instance_label=self._policy_instance_label,
+                resource_context=resource_context,
+            )
+
+            if not decisions["allowed"]:
+                raise AuthorizationError(policy_instance_name=self._policy_instance_name, policy_path=policy_path) # type: ignore[arg-type]
+
+            return handler(*args, **kwargs)
+
+        return cast(Handler, decorated)
+    
+    def check(self, opts: CheckOptions) -> CheckMiddleware:
```

Comment:
I think it would be better to keep `CheckOptions` an internal implementation detail.
The `check` function can take the fields of `CheckOptions` as parameters and construct a an instance internally, including validation that all necessary information is indeed passed in.
It may not seem the most elegant, but Python programmers are pretty used to functions that take lots of optional parameters. This also works better with code editors because the documentation for the function shows up automatically when you call it so coders can see everything they need right in front of them instead of jumping to the definition of the `CheckOptions` class.